File: backend/src/backend/models/question.py
from typing import Any, List, Literal, Optional, Tuple
from pydantic import BaseModel, Field

Subject = Literal["math", "chemistry", "physics", "biology", "computer science"]
Difficulty = Literal["easy","medium","hard"]

class MCQ(BaseModel):
    type: Literal["mcq"]
    choices: List[str]
    answer: str

class TF(BaseModel):
    type: Literal["tf"]
    answer: bool

class Numeric(BaseModel):
    type: Literal["numeric"]
    answer: float

class FIB(BaseModel):
    type: Literal["fib"]
    answer: str

class ShortAnswer(BaseModel):
    type: Literal["short_answer"]
    answer: str

class Matching(BaseModel):
    type: Literal["matching"]
    left: List[str]
    right: List[str]
    answer: List[Tuple]

class Ordering(BaseModel):
    type: Literal["ordering"]
    choices: List[str]
    answer: List[str]

class FR(BaseModel):
    type: Literal["fr"]
    answer: str
    points: int
    rubric: str

class AgentGeneratedQuestion(BaseModel):
    data: MCQ | TF | Numeric | FIB | ShortAnswer | Matching | Ordering | FR
    text: str = Field(description="The question text")
    subject: Subject = Field(description="The subject of the question")
    topic: str = Field(description="The topic of the question")
    source_url: Optional[str] = Field(description="The URL of the source of the question")
    difficulty: Difficulty
    image_url: Optional[str] = Field(description="The URL of the image of the question")

class AgentGeneratedQuestionList(BaseModel):
    questions: List[AgentGeneratedQuestion]

class Question(BaseModel):
    id: str
    question: AgentGeneratedQuestion
    student_answer: Optional[Any] = None
    is_completed: bool = False
    points: int = 1
    points_earned: Optional[int] = None

class FRGrade(BaseModel):
    score: int
    explanation: str

class GradeRequest(BaseModel):
    question: AgentGeneratedQuestion
    student_answer: str

class GradeResponse(BaseModel):
    status: str
    step: str
    message: str
    data: Optional[dict] = None
    error: Optional[str] = None

class AutoGradeRequest(BaseModel):
    question_id: str
    student_answer: str

class AutoGradeResponse(BaseModel):
    is_correct: bool
    points_earned: int
    max_points: int
    explanation: str
    correct_answer: str

class TestResult(BaseModel):
    percentage: float
    total_points: int
    points_earned: int
    summary: str
    improvements: List[str]

File: frontend/src/types/question.ts
export type Subject = 'math' | 'chemistry' | 'physics' | 'biology' | 'computer science';
export type Difficulty = 'easy' | 'medium' | 'hard';

export interface MCQ {
  type: 'mcq';
  choices: string[];
  answer: string;
}

export interface TF {
  type: 'tf';
  answer: boolean;
}

export interface Numeric {
  type: 'numeric';
  answer: number;
}

export interface FIB {
  type: 'fib';
  answer: string;
}

export interface ShortAnswer {
  type: 'short_answer';
  answer: string;
}

export interface Matching {
  type: 'matching';
  left: string[];
  right: string[];
  answer: [number, number][];
}

export interface Ordering {
  type: 'ordering';
  choices: string[];
  answer: string[];
}

export interface FR {
  type: 'fr';
  answer: string;
  points: number;
  rubric: string;
}

export type QuestionData = MCQ | TF | Numeric | FIB | ShortAnswer | Matching | Ordering | FR;

export interface AgentGeneratedQuestion {
  data: QuestionData;
  text: string;
  subject: Subject;
  topic: string;
  source_url?: string;
  difficulty: Difficulty;
  image_url?: string;
}

export interface Question {
  id: string;
  question: AgentGeneratedQuestion;
  student_answer?: any;
  is_completed: boolean;
  points: number;
  points_earned?: number;
}

export interface QuestionWithOptionalCompletion extends Omit<Question, 'is_completed'> {
  is_completed?: boolean;
}

export interface StudentAnswer {
  answer: string | boolean | number | [number, number][] | string[];
}
